### PR TITLE
tor-browser-bundle-bin: 11.5.8 -> 12.0.3

### PR DESCRIPTION
**Pull request by panicgh**
###### Description of changes

https://blog.torproject.org/new-release-tor-browser-120/ (huge list of changes)

https://blog.torproject.org/new-release-tor-browser-1201/

The default bookmarks file seems to have been gotten removed, and `alsa-lib` has to be added unconditionally. Due to the big number of changes, please review carefully. It seems to work fine for me (incl. video/audio), but maybe I overlooked something.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by hax404 — APPROVED** [review]

**Comment by panicgh** [comment]
I experience an issue: on the first start after the previous/old TBB has been run, the new TBB starts just fine. But when I restart the new TBB again, it fails to connect to Tor and shows the all-red error page. This seems repeatable, i.e. I restarted a dozen times and never succeeded to connect to Tor. However, when I start the previous version, close it, then start the new TBB version, it works again for a single time.

**Comment by panicgh** [comment]
Looks like a key in `about:config` is not updated: for example `network.proxy.socks` is set to `file:///run/user/1000/Tor-29/socks.socket`, then I restart TBB, the key still has the same value, but now the socket file doesn't exist any more and I have to manually enter `file:///run/user/1000/Tor-30/socks.socket`. Then everything works fine.

Can anyone confirm this behaviour?

**Comment by mweinelt** [comment]
Yes, I can confirm I see the same issue.

**Comment by panicgh** [comment]
Rebased and updated to 12.0.1.

However, the "proxy error" right after startup persists. The current workaround is to set `network.proxy.socks` in `about:config` to the right path.

**Comment by panicgh** [comment]
Rebased, then updated to 12.0.2. The problem with `network.proxy.socks` persists.

https://blog.torproject.org/new-release-tor-browser-1202/

**Comment by panicgh** [comment]
* Rebased to latest `master`
* Bump version to 12.0.3

The problem with `network.proxy.socks` is still there. I don't know how to fix it.

I use this version for some time now and on every start fix `network.proxy.socks` in `about:config` by hand. Should this PR be merged regardless the issue in order to not use the outdated/insecure TBB and request from users to fix the `network.proxy.socks` setting, like I do?

Or should the TBB package be marked as broken or insecure instead?

**Comment by mweinelt** [comment]
Can you create an upstream issue about this?

**Comment by SuperSandro2000** [comment]
I did a short investigation of this and concluded the following:

- the socket directory in /run/user/1000/Tor is not cleaned up on shutdown
- on startup a - with a counting up number is appended and written into the extensions.torlauncher.tmp_ipc_dir setting through I guess https://gitlab.torproject.org/tpo/applications/tor-browser/-/blob/tor-browser-102.8.0esr-12.5-1/toolkit/components/tor-launcher/TorLauncherUtil.jsm#L118
- the network.proxy.socks always points at /run/user/1000/Tor/ and the browser gets very confused about that


https://matrix.to/#/!NBBFPbiuttRgTqbrcY:nixos.org/$i9ZvKrwVRojYQ6s5AMSRGzphyo_wBp5aE_XTEhqnADs?via=nixos.org&via=matrix.org&via=nixos.dev

---

see https://github.com/SuperSandro2000/nixpkgs/commit/77342ddbfabbcbdc6a8eb7fc60752cd5f12bdf70 for a potential fix

**Comment by panicgh** [comment]
Thank you very much for the workaround. I incorporated it and not TBB runs even after restarts!
I would suggest to accept this PR as it is with the workaround, so that TBB can be updated and is usable again. If later a better approach can be found, it can be changed/improved then.

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix [hunk 1]
```diff
@@ -325,6 +323,9 @@ stdenv.mkDerivation rec {
     # chance that TBB would continue using old font files.
     rm -rf "\$HOME/.cache/fontconfig"
 
+    # Workaround a bug in 12.0.X that Tor directories are not cleaned up and tor gets confused where its socket is
```
Comment (dotlambda) [inline pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix:326]: link to upstream issue

File: pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix [hunk 2]
```diff
@@ -107,7 +109,7 @@ let
         "https://tor.eff.org/dist/torbrowser/${version}/tor-browser-linux32-${version}_${lang}.tar.xz"
         "https://tor.calyxinstitute.org/dist/torbrowser/${version}/tor-browser-linux32-${version}_${lang}.tar.xz"
       ];
-      sha256 = "sha256-TGdJ5yIeo0YQ4XSsb9lv3vuW6qEjhFe7KBmkjYO6fAc=";
+      sha256 = "sha256-t1tnEZtiRig2r8GNJpqT+J0XoxCLMyUsI9tX6aa0lYk=";
```
Comment (dotlambda) [inline pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix:112]:
```suggestion
      hash = "sha256-t1tnEZtiRig2r8GNJpqT+J0XoxCLMyUsI9tX6aa0lYk=";
```

File: pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix [hunk 3]
```diff
@@ -97,7 +99,7 @@ let
         "https://tor.eff.org/dist/torbrowser/${version}/tor-browser-linux64-${version}_${lang}.tar.xz"
         "https://tor.calyxinstitute.org/dist/torbrowser/${version}/tor-browser-linux64-${version}_${lang}.tar.xz"
       ];
-      sha256 = "sha256-/KK9oTijk5dEziAwp5966NaM2V4k1mtBjTJq88Ct7N0=";
+      sha256 = "sha256-bOGY/RdwD6O7QIuOiBw7OVnZfpumGGso6hwMJJwN2g0=";
```
Comment (dotlambda) [inline pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix:102]:
```suggestion
      hash = "sha256-bOGY/RdwD6O7QIuOiBw7OVnZfpumGGso6hwMJJwN2g0=";
```

File: pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix (hunk 1 above)
Comment (panicgh) [reply pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix:326]: Currently, there isn't any.

Comment (dotlambda) [reply pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix:326]: Can we open one?

**Comment by a12l** [comment]
@panicgh and @SuperSandro2000: Pinging you here also because GH notifications is a trash fire. The workaround doesn't work for me and others. #219750

**Comment by SuperSandro2000** [comment]
> because GH notifications is a trash fire.

Not really, just really full for me 😜

**Pull request merged**